front_end/dart_tests/test_helpers.dart

```diff
@@ -0,0 +1,273 @@
+  rawLocationToUILocation(_) => _location;
+  FakeChromeBindings({String currentFileUrl}) {
+    _location._url = currentFileUrl;
+  }
+  set currentFileUrl(x) => _location._url = x;
+}
+
+/// A fake for devtools UILocation.
+///
+/// Some of the things we provide really belong on other objects,
+/// but we add them here to avoid having quite so many fakes.
+class UILocation {
+  get columnNumber => 7;
+  get lineNumber => 42;
+
+  /// Just use this as its own uiSourceCode.
+  get uiSourceCode => this;
+  // url would really be on UISourceCode.
+  var _url;
+  url() => _url;
+}
+
+/// A fake for the devtools SDK.DebuggerModel.CallFrame.
+class FakeCallFrame {
+  functionLocation() => this;
+  script() => this;
+  get sourceURL => 'someURL';
+
+  /// If true we retain scopeChainForWeb, otherwise scopeChainForLib.
+  bool inWeb;
+  scopeChain() => inWeb ? scopeChainWeb() : scopeChainLib();
+
+  /// Allow creating a call frame in one of two flavors, one hard-coded to a
+  /// particular file in the web directory, and another in lib.
+  FakeCallFrame(this.inWeb);
+  FakeCallFrame.inWeb() : this.inWeb = true;
+  FakeCallFrame.inLib() : this.inWeb = false;
+
+  /// We don't actually evaluate, the test sets the desired result.
+  var evaluationResult;
+  Promise evaluate(evaluateOptions, userGesture, awaitPromise) {
+    return futureToPromise(Future.value(evaluationResult));
+  }
+
+  /// A hard-coded scope chain for a sample file in the web directory.
+  scopeChainWeb() => [
+        Bindings('self', {'a': 1, 'b': 2, 'c': 3}),
+        Bindings('library',
+            {'foo__bar__web__main': _libraryBindingsWeb, 'gets_ignored': 4}),
+        Bindings('global_ignored', {})
+      ];
+
+  /// Bindings for variables in a library in /web
+  get _libraryBindingsWeb =>
+      Bindings('foo__bar__web__main', {'_private': 'lib_prop'});
+
+  /// A hard-coded scope chain for a sample file in the lib directory.
+  scopeChainLib() => [
+        Bindings('self', {'x': 99}),
+        Bindings('library',
+            {'foo__bar__library': _libraryBindingsLib, 'gets_ignored': 4}),
+        Bindings('global_ignored', {})
+      ];
+
+  /// Bindings for variables in a library in /lib.
+  get _libraryBindingsLib =>
+      Bindings('foo__bar__library', {'_lib_var': 'lib_prop'});
+  get debuggerModel => this;
+  runtimeModel() => this;
+  releaseObjectGroup(_) => {};
+  // Function names are either of this form (for unnamed things) or the actual
+  // name (for named functions).
+  get _webFunctionName => 'foo__bar__web__main';
+  get _libFunctionName => 'foo__bar__library';
+  get _functionName => inWeb ? _webFunctionName : _libFunctionName;
+  get functionName => 'dart_library.library.$_functionName.actualName';
+}
+
+/// Fake for a devtools "Bindings". This is not the same as the Bindings
+/// namespace, it's an anonymous object with a name and properties, where the
+/// properties are SDK.RemoteObjectProperties.
+class Bindings {
```

Comment:
can you provide a few more types for clarity for this code. E.g. what is the type of _name?

Reply:
Done.